extensions/gitpod/src/authentication.ts

```diff
@@ -0,0 +1,272 @@
+import { GitpodServer } from './gitpodServer';
+import Log from './common/logger';
+import { arrayEquals } from './common/utils';
+import { Disposable } from './common/dispose';
+import TelemetryReporter from './telemetryReporter';
+
+interface SessionData {
+	id: string;
+	account?: {
+		label?: string;
+		displayName?: string;
+		id: string;
+	};
+	scopes: string[];
+	accessToken: string;
+}
+
+export default class GitpodAuthenticationProvider extends Disposable implements vscode.AuthenticationProvider {
+	private _sessionChangeEmitter = new vscode.EventEmitter<vscode.AuthenticationProviderAuthenticationSessionsChangeEvent>();
+	private _logger: Log;
+	private _telemetry: TelemetryReporter;
+	private _gitpodServer: GitpodServer;
+	private _keychain: Keychain;
+
+	private _sessionsPromise: Promise<vscode.AuthenticationSession[]>;
+
+	constructor(private readonly context: vscode.ExtensionContext, logger: Log, telemetry: TelemetryReporter) {
+		super();
+
+		this._logger = logger;
+		this._telemetry = telemetry;
+
+		const gitpodHost = vscode.workspace.getConfiguration('gitpod').get<string>('host')!;
+		const serviceUrl = new URL(gitpodHost);
+		this._gitpodServer = new GitpodServer(serviceUrl.toString(), this._logger);
+		this._keychain = new Keychain(this.context, `gitpod.auth.${serviceUrl.hostname}`, this._logger);
+		this._register(vscode.workspace.onDidChangeConfiguration(e => {
+			if (e.affectsConfiguration('gitpod.host')) {
+				const gitpodHost = vscode.workspace.getConfiguration('gitpod').get<string>('host')!;
+				const serviceUrl = new URL(gitpodHost);
+				this._gitpodServer.dispose();
+				this._gitpodServer = new GitpodServer(serviceUrl.toString(), this._logger);
+				this._keychain = new Keychain(this.context, `gitpod.auth.${serviceUrl.hostname}`, this._logger);
+
+				this.checkForUpdates();
+			}
+		}));
+
+		// Contains the current state of the sessions we have available.
+		this._sessionsPromise = this.readSessions();
+
+		this._register(vscode.authentication.registerAuthenticationProvider('gitpod', 'Gitpod', this, { supportsMultipleAccounts: false }));
+		this._register(this.context.secrets.onDidChange(() => this.checkForUpdates()));
+	}
+
+	get onDidChangeSessions() {
+		return this._sessionChangeEmitter.event;
+	}
+
+	async getSessions(scopes?: string[]): Promise<vscode.AuthenticationSession[]> {
+		// For the Gitpod scope list, order doesn't matter so we immediately sort the scopes
+		const sortedScopes = scopes?.sort() || [];
+		this._logger.info(`Getting sessions for ${sortedScopes.length ? sortedScopes.join(',') : 'all scopes'}...`);
+		const sessions = await this._sessionsPromise;
+		const finalSessions = sortedScopes.length
+			? sessions.filter(session => arrayEquals([...session.scopes].sort(), sortedScopes))
+			: sessions;
+
+		this._logger.info(`Got ${finalSessions.length} sessions for ${sortedScopes?.join(',') ?? 'all scopes'}...`);
+		return finalSessions;
+	}
+
+	private async checkForUpdates() {
```

Comment:
probably it is minor but this code can race, it would be better to make it cancellabe

Reply:
in theory yes, but I think it's really hard or impossible to happen as it depends on user input and the `this._sessionsPromise` is most probably already resolved, making it cancellable will make it race too in the second await, I think a queue will "fix" it, but I'll leave it for later